fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,224 @@
+        x_norm = (x**2).sum(1).view(-1, 1)
+        if y is not None:
+            y_norm = (y**2).sum(1).view(1, -1)
+        else:
+            y = x
+            y_norm = x_norm.view(1, -1)
+        Pdist = x_norm + y_norm - 2.0 * torch.mm(x, torch.transpose(y, 0, 1))
+        Pdist[Pdist < 0] = 0
+        return Pdist
+
+    def h1_mean_var_gram(
+        self, Kx: torch.Tensor, Ky: torch.Tensor, Kxy: torch.Tensor, is_var_computed: bool, use_1sample_U: bool = True
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]:
+        """compute value of MMD and std of MMD using kernel matrix."""
+        Kxxy = torch.cat((Kx, Kxy), 1)
+        Kyxy = torch.cat((Kxy.transpose(0, 1), Ky), 1)
+        Kxyxy = torch.cat((Kxxy, Kyxy), 0)
+        nx = Kx.shape[0]
+        ny = Ky.shape[0]
+        is_unbiased = True
+        if is_unbiased:
+            xx = torch.div((torch.sum(Kx) - torch.sum(torch.diag(Kx))), (nx * (nx - 1)))
+            yy = torch.div((torch.sum(Ky) - torch.sum(torch.diag(Ky))), (ny * (ny - 1)))
+            # one-sample U-statistic.
+            if use_1sample_U:
+                xy = torch.div((torch.sum(Kxy) - torch.sum(torch.diag(Kxy))), (nx * (ny - 1)))
+            else:
+                xy = torch.div(torch.sum(Kxy), (nx * ny))
+            mmd2 = xx - 2 * xy + yy
+        else:
+            xx = torch.div((torch.sum(Kx)), (nx * nx))
+            yy = torch.div((torch.sum(Ky)), (ny * ny))
+            # one-sample U-statistic.
+            if use_1sample_U:
+                xy = torch.div((torch.sum(Kxy)), (nx * ny))
+            else:
+                xy = torch.div(torch.sum(Kxy), (nx * ny))
+            mmd2 = xx - 2 * xy + yy
+        if not is_var_computed:
+            return mmd2, None, Kxyxy
+        hh = Kx + Ky - Kxy - Kxy.transpose(0, 1)
+        V1 = torch.dot(hh.sum(1) / ny, hh.sum(1) / ny) / ny
+        V2 = (hh).sum() / (nx) / nx
+        varEst = 4 * (V1 - V2**2)
+        return mmd2, varEst, Kxyxy
+
+    def MMDu(
+        self,
+        Fea: torch.Tensor,
+        len_s: int,
+        Fea_org: torch.Tensor,
+        sigma: torch.Tensor,
+        sigma0: torch.Tensor,
+        epsilon: torch.Tensor,
+        is_smooth: bool = True,
+        is_var_computed: bool = True,
+        use_1sample_U: bool = True,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]:
+        """compute value of deep-kernel MMD and std of deep-kernel MMD using merged data."""
+        X = Fea[0:len_s, :]  # fetch the sample 1 (features of deep networks)
+        Y = Fea[len_s:, :]  # fetch the sample 2 (features of deep networks)
+        X_org = Fea_org[0:len_s, :]  # fetch the original sample 1
+        Y_org = Fea_org[len_s:, :]  # fetch the original sample 2
+        L = 1  # generalized Gaussian (if L>1)
+        Dxx = self.Pdist2(X, X)
+        Dyy = self.Pdist2(Y, Y)
+        Dxy = self.Pdist2(X, Y)
+        Dxx_org = self.Pdist2(X_org, X_org)
```

Comment:
I think we can put these inside the `if is_smooth` block, since they only need to be computed if that is true. Otherwise they aren't used.

Reply:
Maybe also rename these to expand `org` to `original` just to be explicit